engine/src/setup.ts

```diff
@@ -0,0 +1,259 @@
+import assert from "assert";
+import AvailableCommand from "./available-command";
+import Engine from "./engine";
+import {
+  AdvTechTile,
+  AdvTechTilePos,
+  Booster,
+  Command,
+  Federation,
+  FinalTile,
+  Player,
+  ScoringTile,
+  TechTile,
+  TechTilePos,
+} from "./enums";
+import SpaceMap, { MapTile } from "./map";
+
+export enum SetupType {
+  Booster = "booster",
+  TechTile = "techTile",
+  AdvTechTile = "advTechTile",
+  TerraformingFederation = "terraformingFederation",
+  RoundScoringTile = "roundScoringTile",
+  FinalScoringTile = "finalScoringTile",
+  MapTile = "mapTile",
+}
+
+export type SetupPosition = number | TechTilePos | AdvTechTilePos;
+export type SetupOption = Booster | TechTile | AdvTechTile | Federation | ScoringTile | FinalTile | string; //string is for MapTile name
+
+type SetupFactoryOption = {
+  position: SetupPosition;
+  options: SetupOption[];
+};
+
+export type AvailableSetupOption = { type: SetupType } & SetupFactoryOption;
+
+type SetupFactory = {
+  type: SetupType;
+  nextAvailable: () => SetupFactoryOption | null;
+  applyOption: (option: SetupOption, position: SetupPosition) => void;
+  init: () => void;
+};
+
+function techFactory(
+  engine: Engine,
+  type: SetupType,
+  tilePos: (TechTilePos | AdvTechTilePos)[],
+  techTiles: (TechTile | AdvTechTile)[],
+  count: number
+) {
+  return {
+    type: type,
```

Comment:
Here (and below) you can just do:
```suggestion
    type,
```